### PR TITLE
feat(harness): self-improving runtime, plan lifecycle, evolve command (v5.0.1)

### PR DESCRIPTION
**Pull request by donovan-yohan**
## Summary
- Add self-improving harness runtime with `.harness/` scaffold, persistence scripts, and run-state protocol
- Add `/harness:evolve` command for self-modification proposals via harness-evolver agent
- Wire evolve into reflect (Phase 5.8) and evolution summary into complete (Phase 5.7)
- Extend `/harness:review` with evaluator pass and structured output
- Extend `/harness:bug` with retroactive harness trace (fitness function)
- Add `/harness:init` runtime scaffold, runtime-spec reference, learnings format improvements
- Deprecate `/harness:loop` in favor of run-state protocol
- Add deterministic persistence scripts with test suite
- Harden adversarial-review.sh (remove deprecated `--bare` flag)
- Bump harness plugin version to 5.0.1

## Test plan
- [ ] `go test ./...` passes (version sync)
- [ ] `go build -o belayer ./cmd/belayer` builds clean
- [ ] `plugins/harness/scripts/test-harness-scripts.sh` passes all persistence script tests
- [ ] Manual: `/harness:init` scaffolds `.harness/` directory
- [ ] Manual: `/harness:evolve` produces evolution proposals

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: plugins/harness/scripts/harness-read-state.sh [hunk 1]
```diff
@@ -0,0 +1,22 @@
+#!/usr/bin/env bash
+# Read run-state.json to stdout. Empty output if file doesn't exist.
```
Comment (donovan-yohan) [inline plugins/harness/scripts/harness-read-state.sh]: **[comment-analyzer]** Header comment says "Empty output if file doesn't exist" but line 21 actually returns `{"phase": "none"}` — not empty output. The test suite also asserts the sentinel JSON, confirming the implementation intent.

Suggestion: `# Read run-state.json to stdout. Returns {"phase": "none"} if file doesn't exist.`

File: plugins/harness/scripts/harness-read-state.sh [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    --harness-dir) HARNESS_DIR="$2"; shift 2 ;;
+    *) echo "Usage: $0 --harness-dir <path>" >&2; exit 1 ;;
+  esac
+done
+
+[ -z "$HARNESS_DIR" ] && { echo "Error: --harness-dir required" >&2; exit 1; }
+
+STATE_FILE="$HARNESS_DIR/run-state.json"
+
+if [ -f "$STATE_FILE" ]; then
```
Comment (donovan-yohan) [inline plugins/harness/scripts/harness-read-state.sh:18]: **[silent-failure-hunter]** HIGH: If `run-state.json` exists but contains corrupt/invalid JSON (truncated write, disk failure, concurrent worktree access), this outputs garbage to stdout with exit code 0. Every consumer will fail with obscure Python JSON parse errors that don't point back to the corrupt file.

Suggestion: Add JSON validation before outputting:
```bash
if [ -f "$STATE_FILE" ]; then
  if python3 -c "import json,sys; json.load(sys.stdin)" < "$STATE_FILE" 2>/dev/null; then
    cat "$STATE_FILE"
  else
    echo "Error: $STATE_FILE contains invalid JSON" >&2
    exit 1
  fi
else
  echo '{"phase": "none"}'
fi
```

File: plugins/harness/scripts/harness-update-state.sh [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        'plan': plan,
+        'design_doc': design_doc,
+        'branch': branch,
+        'phase': phase,
+        'completed_phases': [{'name': phase, 'completed_at': now}],
+        'started_at': now,
+        'last_updated': now,
+    }
+
+# Non-atomic write: write to temp file then rename to avoid truncation on failure
```
Comment (donovan-yohan) [inline plugins/harness/scripts/harness-update-state.sh]: **[comment-analyzer]** Comment says "Non-atomic write" but the write-then-rename pattern (`os.rename(tmp_file, state_file)`) IS the standard POSIX atomic write pattern. The comment contradicts itself — says "non-atomic" then describes the exact technique for atomicity. Same issue in `harness-write-metrics.sh:144`.

Suggestion: `# Atomic write: write to temp file then rename to avoid truncation on failure`

File: plugins/harness/scripts/harness-write-run.sh [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+        print(f"Error: --data-file '{data_file}' is not valid JSON: {e}", file=sys.stderr)
+        sys.exit(1)
+
+record = {
+    'phase': os.environ['PHASE'],
+    'branch': os.environ.get('BRANCH', 'unknown'),
+    'timestamp': os.environ['NOW'],
+    'data': data
+}
+with open(os.environ['RUN_FILE'], 'w') as f:
```
Comment (donovan-yohan) [inline plugins/harness/scripts/harness-write-run.sh]: **[silent-failure-hunter]** HIGH: This writes directly with `open()` + `json.dump()` without error handling, unlike sibling scripts (`harness-update-state.sh`, `harness-write-metrics.sh`) which both use temp-file-then-rename with `OSError` handling. A write failure (disk full, permissions) produces an unhandled Python traceback and could leave a partially-written file. A missing run record silently loses historical data that `/harness:bug` retroactive trace depends on.

Suggestion: Add the same temp-file-then-rename pattern:
```python
tmp_file = os.environ['RUN_FILE'] + '.tmp'
try:
    with open(tmp_file, 'w') as f:
        json.dump(record, f, indent=2)
    os.rename(tmp_file, os.environ['RUN_FILE'])
except OSError as e:
    print(f"Error: failed to write run record: {e}", file=sys.stderr)
    try: os.unlink(tmp_file)
    except OSError: pass
    sys.exit(1)
```

File: plugins/harness/scripts/harness-write-run.sh [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+[ -z "$HARNESS_DIR" ] && { echo "Error: --harness-dir required" >&2; exit 1; }
+[ -z "$PHASE" ] && { echo "Error: --phase required" >&2; exit 1; }
+
+RUNS_DIR="$HARNESS_DIR/runs"
+mkdir -p "$RUNS_DIR"
+
+# Compute timestamp once to ensure TIMESTAMP and NOW refer to the same instant
+_TS_RAW=$(date -u +%Y-%m-%dT%H%M%SZ)
+TIMESTAMP="$_TS_RAW"
+NOW=$(echo "$_TS_RAW" | sed 's/T\([0-9]\{2\}\)\([0-9]\{2\}\)\([0-9]\{2\}\)Z$/T\1:\2:\3Z/')
```
Comment (donovan-yohan) [inline plugins/harness/scripts/harness-write-run.sh]: **[code-simplifier]** The sed-based timestamp reformatting is fragile. Both formats can be computed directly from `date`:
```bash
TIMESTAMP=$(date -u +%%Y-%%m-%%dT%%H%%M%%SZ)
NOW=$(date -u +%%Y-%%m-%%dT%%H:%%M:%%SZ)
```
Two `date` calls in the same script will produce the same second in practice. If exact same-instant is critical, capture epoch once and format twice.

File: plugins/harness/scripts/harness-write-metrics.sh [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+except OSError as e:
+    print(f"Error: failed to read {metrics_file}: {e}", file=sys.stderr)
+    sys.exit(1)
+
+if metric_type == 'review-effectiveness':
+    agent_name = os.environ['AGENT']
+    agents = data.setdefault('agents', {})
+    agent = agents.setdefault(agent_name, {
+        'runs': 0, 'findings': 0, 'false_positives': 0,
+        'unique_catches': 0, 'last_run': None, 'disabled': False, 'disable_reason': None
```
Comment (donovan-yohan) [inline plugins/harness/scripts/harness-write-metrics.sh]: **[type-design-analyzer]** `disabled: False` and `disable_reason: None` are ghost fields — created here via `setdefault` but never written to by any script, and explicitly noted as "not yet implemented" in `runtime-spec.md`. This creates schema inconsistency: agents from `harness-write-metrics.sh` have these fields, but agents from `harness-init-runtime.sh` don't.

Suggestion: Remove `'disabled': False, 'disable_reason': None` from the `setdefault` dict until the feature is built.

File: plugins/harness/scripts/harness-init-runtime.sh [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+if [ -f "$HARNESS_DIR/manifest.yaml" ] && [ "$FORCE" != "true" ]; then
+  echo "Warning: $HARNESS_DIR already initialized (use --force to overwrite)" >&2
+  exit 1
+fi
+
+# If --force and manifest exists, back up existing config before overwriting
+if [ "$FORCE" = "true" ] && [ -f "$HARNESS_DIR/manifest.yaml" ]; then
+  BACKUP_TS=$(date -u +%Y%m%dT%H%M%SZ)
+  cp "$HARNESS_DIR/manifest.yaml" "$HARNESS_DIR/manifest.yaml.bak.$BACKUP_TS"
+  [ -f "$HARNESS_DIR/config.yaml" ] && cp "$HARNESS_DIR/config.yaml" "$HARNESS_DIR/config.yaml.bak.$BACKUP_TS"
```
Comment (donovan-yohan) [inline plugins/harness/scripts/harness-init-runtime.sh]: **[silent-failure-hunter]** Under `set -e`, `[ -f "$HARNESS_DIR/config.yaml" ] && cp ...` evaluates to false (exit 1) when config.yaml doesn't exist, which triggers `set -e` to abort the script silently. A user running `--force` on a `.harness/` with `manifest.yaml` but no `config.yaml` gets a silent exit with no error message.

Fix: `[ -f "$HARNESS_DIR/config.yaml" ] && cp "$HARNESS_DIR/config.yaml" "$HARNESS_DIR/config.yaml.bak.$BACKUP_TS" || true`

File: plugins/harness/scripts/harness-init-runtime.sh [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+# Harness Improvements
+
+Audit trail of self-modifications to agent definitions. Append-only.
+
+---
+EOF
+
+# .gitignore
+cat > "$HARNESS_DIR/.gitignore" <<'EOF'
+handoffs/
```
Comment (donovan-yohan) [inline plugins/harness/scripts/harness-init-runtime.sh:121]: **[code-reviewer]** `evolve.md` Phase 4 creates `$HARNESS_DIR/metrics/pre-change-snapshot.json` as an ephemeral per-session file that gets "overwritten each evolve session." This file is not listed in the `.gitignore` template here, so it will be committed to the repo.

Suggestion: Add `metrics/pre-change-snapshot.json` to this `.gitignore` template.

File: plugins/harness/scripts/harness-resolve-dir.sh [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+
+REPO_ROOT="."
+while [[ $# -gt 0 ]]; do
+  case $1 in
+    --repo-root) REPO_ROOT="$2"; shift 2 ;;
+    *) echo "Usage: $0 [--repo-root <path>]" >&2; exit 1 ;;
+  esac
+done
+
+REPO_SLUG=$(basename "$(cd "$REPO_ROOT" && git rev-parse --show-toplevel 2>/dev/null || echo "$REPO_ROOT")")
```
Comment (donovan-yohan) [inline plugins/harness/scripts/harness-resolve-dir.sh:20]: **[silent-failure-hunter]** If `git` is not installed, `git rev-parse` silently falls through to `echo "$REPO_ROOT"`, making `REPO_SLUG` the basename of whatever path was passed (e.g., `.` produces `REPO_SLUG="."`). The global path becomes `~/.harness/.` which corrupts the lookup. Also, if `$REPO_ROOT` doesn't exist, `cd` fails unpredictably.

Suggestion:
```bash
if [ ! -d "$REPO_ROOT" ]; then
  echo "Error: --repo-root '$REPO_ROOT' is not a directory" >&2
  exit 1
fi
```

File: plugins/harness/scripts/harness-write-proposal.sh [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+  echo ""
+  echo "- **Date:** $DATE"
+  echo "- **Signal:** $SIGNAL"
+  echo "- **Agent:** $AGENT"
+  echo "- **Scope:** $SCOPE"
+  echo "- **Status:** pending"
+  echo ""
+  echo "## Current"
+  echo ""
+  if [ -f "$CURRENT_FILE" ]; then
```
Comment (donovan-yohan) [inline plugins/harness/scripts/harness-write-proposal.sh:54]: **[silent-failure-hunter]** When `--current-file` is explicitly provided but the file doesn't exist (temp cleaned up, path typo), this silently falls back to placeholder text. The caller intended to provide content but gets "(no current text provided)" with no warning. Same for `--proposed-file` and `--reasoning-file`.

Suggestion: Warn when an explicitly-provided path is missing:
```bash
if [ -n "$CURRENT_FILE" ] && [ ! -f "$CURRENT_FILE" ]; then
  echo "Warning: --current-file '$CURRENT_FILE' not found, using placeholder" >&2
fi
```

File: plugins/harness/commands/evolve.md [hunk 11]
```diff
@@ -0,0 +1,237 @@
...
+8. Update `$HARNESS_DIR/metrics/learning-efficacy.json`:
+   - For each new learning from this session, check if any bugs in this session match existing learning categories (recurrence detection)
+   - If a learning's recommendation was supposed to prevent this class of bug but the bug occurred anyway, increment `recurrence_count`
+   - If a learning's recommendation was consulted and the relevant bug class did NOT occur, increment `prevented_count`
+   - Use the metrics script for each learning:
+     ```bash
+     bash ${CLAUDE_PLUGIN_ROOT}/scripts/harness-write-metrics.sh \
+       --harness-dir "$HARNESS_DIR" \
+       --metric learning-efficacy \
+       --plan-slug "{learning-id}" \
```
Comment (donovan-yohan) [inline plugins/harness/commands/evolve.md]: **[comment-analyzer]** This instructs the LLM to use `--plan-slug`, `--drift`, `--surprises` (backward-compat names) for `learning-efficacy` metrics, but the same PR introduces dedicated `--learning-id`, `--recurrence`, `--prevented` parameters.

Suggestion: Use the canonical parameter names:
```
--learning-id "{learning-id}" \
--recurrence {recurrence-increment} \
--prevented {prevented-increment}
```
Same fix needed in `skills/harness-evolve/SKILL.md`.

File: plugins/harness/commands/review.md [hunk 12]
```diff
@@ -224,6 +245,52 @@ Context-isolated adversarial review using the bundled `scripts/adversarial-revie
...
+27. **Update run-state** (if `.harness/` runtime exists, reuse `$HARNESS_DIR` from step 24):
+    ```bash
+    [ -n "$HARNESS_DIR" ] && bash ${CLAUDE_PLUGIN_ROOT}/scripts/harness-update-state.sh \
+      --harness-dir "$HARNESS_DIR" \
+      --phase "review"
+    ```
+
 ### Report
 
 23. Output:
```
Comment (donovan-yohan) [inline plugins/harness/commands/review.md]: **[comment-analyzer]** Step numbering is non-sequential: Phase 8 contains steps 24-27 (lines 250-295), but this Report section is labeled step 23 despite appearing AFTER them. An LLM following these instructions top-to-bottom hits step 22 then 24 then 25 then 26 then 27 then 23.

Suggestion: Renumber to step 28 (or Phase 9). Same fix needed in `skills/harness-review/SKILL.md`.

**Review by donovan-yohan — COMMENTED** [review]
## PR Review Summary

**Agents run:** code-reviewer, silent-failure-hunter, pr-test-analyzer, type-design-analyzer, comment-analyzer, code-simplifier

### Overall Assessment
The shell scripts are well-structured with `set -euo pipefail`, proper argument validation, and atomic writes via temp-file-then-rename. The Go changes are clean with three-way version sync (plugin.json, registry.go, test). Test coverage is solid (58/58 pass). The weakest link is boundary gaps between deterministic and LLM operations, and several stale references from the `loop.md` deletion.

### Key Themes
- **2 HIGH severity**: Corrupt JSON passthrough in read-state, inconsistent write pattern in write-run
- **Stale `/harness:loop` references** in 6+ files not modified by this PR (refactor.md, refactor-status.md, prune.md + skill mirrors) — command was deleted but references remain as "Next Steps" directives
- **Comment/spec inaccuracies**: "Non-atomic write" comments on atomic operations, header comment contradicting code behavior, spec/implementation schema drift
- **Ghost fields**: `disabled`/`disable_reason` written by code but absent from spec and not yet implemented
- **Dead code**: `features.evolve`/`features.test` flags in manifest.yaml are never read by any command or script

### Files NOT in diff with stale `/harness:loop` references (cannot comment inline)
- `plugins/harness/commands/refactor.md` (lines 116-123)
- `plugins/harness/commands/refactor-status.md` (lines 70-71)
- `plugins/harness/commands/prune.md` (line 98)
- `plugins/harness/agents/harness-pruner.md` (line 233)
- `skills/harness-refactor/SKILL.md`, `skills/harness-refactor-status/SKILL.md`, `skills/harness-prune/SKILL.md`

### Agents with no critical findings
- **pr-test-analyzer**: 58 tests all pass; identified missing edge case tests (invalid phase rejection, corrupt JSON recovery, scope validation) but existing coverage is solid
- **type-design-analyzer**: `harness-resolve-dir.sh` rated highest (8/10 encapsulation); `Registry` Go type is clean; learning entry format rated lowest (2/10 enforcement) due to pure LLM convention

### Test Coverage Gaps Identified
- No test for invalid phase rejection in `harness-update-state.sh`
- No test for duplicate phase re-entry (dedup path)
- No test for corrupt JSON recovery in `harness-read-state.sh`/`harness-write-metrics.sh`
- No test for `--force` with missing `config.yaml`

File: plugins/harness/references/runtime-spec.md [hunk 13]
```diff
@@ -0,0 +1,257 @@
...
+      "last_run": null
+    }
+  },
+  "last_updated": null
+}
+```
+
+<!-- NOTE: Agent disabling (disabled/disable_reason fields) is not yet implemented.
+     Add those fields when the disable-agent feature is built. -->
+
```
[the remark this answers is not in the document]
Comment (donovan-yohan) [reply plugins/harness/references/runtime-spec.md]: Fixed. Removed ghost fields from `harness-write-metrics.sh` and updated the spec NOTE to clarify both spec and script should be updated together when the feature is built. See 5f042f0.

File: plugins/harness/commands/reflect.md [hunk 14]
```diff
@@ -179,6 +179,28 @@ This phase feeds the self-learning review system. When bugs escape `/harness:rev
...
+14.12. Resolve the harness runtime directory:
+    ```bash
+    HARNESS_DIR=$(bash ${CLAUDE_PLUGIN_ROOT}/scripts/harness-resolve-dir.sh --repo-root .)
+    ```
+
+14.13. If `HARNESS_DIR` is not empty:
+    - Invoke `/harness:evolve` using the Skill tool: `Skill("harness:evolve")`. Follow the loaded skill's full process.
+
+    <MANDATORY>
+    You MUST use the Skill tool to invoke `/harness:evolve`. Do NOT classify learnings or generate proposals inline — the evolve command has the persistence script integration, evolver agent dispatch, and auto-apply safety checks that prevent both under-evolution and over-evolution.
```
[the remark this answers is not in the document]
Comment (donovan-yohan) [reply plugins/harness/commands/reflect.md]: Fixed. Changed to invoke `/harness:evolve` as a command, not via `Skill()`. Applied to both `reflect.md` and `harness-reflect/SKILL.md`. See 5f042f0.

File: skills/harness-plan/SKILL.md [hunk 15]
```diff
@@ -108,7 +108,17 @@ harness-plan docs/design-docs/{file}.md         # Plan from specific design doc
...
 
-6. Update `docs/PLANS.md` — add the new plan to the Active Plans table.
+5.5. **Update run-state** (if `.harness/` runtime exists):
+    ```bash
+    HARNESS_DIR=$(bash ${CLAUDE_PLUGIN_ROOT}/scripts/harness-resolve-dir.sh --repo-root .)
+    [ -n "$HARNESS_DIR" ] && bash ${CLAUDE_PLUGIN_ROOT}/scripts/harness-update-state.sh \
+      --harness-dir "$HARNESS_DIR" \
+      --phase "plan" \
+      --plan "docs/exec-plans/active/{filename}" \
+      --design-doc "{design-doc-path}"
```
[the remark this answers is not in the document]
Comment (donovan-yohan) [reply skills/harness-plan/SKILL.md:118]: Fixed. Added `.md` extension to `--plan` path. See 5f042f0.

File: plugins/harness/commands/plan.md [hunk 16]
```diff
@@ -104,7 +104,17 @@ Create a living execution plan from a design document, saved as a versioned arti
...
    ```
 
-6. Update `docs/PLANS.md` — add the new plan to the Active Plans table.
+5.5. **Update run-state** (if `.harness/` runtime exists):
+    ```bash
+    HARNESS_DIR=$(bash ${CLAUDE_PLUGIN_ROOT}/scripts/harness-resolve-dir.sh --repo-root .)
+    [ -n "$HARNESS_DIR" ] && bash ${CLAUDE_PLUGIN_ROOT}/scripts/harness-update-state.sh \
+      --harness-dir "$HARNESS_DIR" \
+      --phase "plan" \
+      --plan "docs/exec-plans/active/{filename}" \
```
[the remark this answers is not in the document]
Comment (donovan-yohan) [reply plugins/harness/commands/plan.md]: Fixed. Added `.md` extension to `--plan` path. See 5f042f0.

File: skills/harness-brainstorm/SKILL.md [hunk 17]
```diff
@@ -56,15 +56,26 @@ harness-brainstorm add user auth      # Brainstorm with initial topic
...
    - Add to the "Key Decisions" table if a non-trivial decision was made
 
+5.5. **Update run-state** (if `.harness/` runtime exists):
+    ```bash
+    HARNESS_DIR=$(bash ${CLAUDE_PLUGIN_ROOT}/scripts/harness-resolve-dir.sh --repo-root .)
+    [ -n "$HARNESS_DIR" ] && bash ${CLAUDE_PLUGIN_ROOT}/scripts/harness-update-state.sh \
+      --harness-dir "$HARNESS_DIR" \
+      --phase "brainstorm" \
+      --design-doc "docs/design-docs/{filename}" \
+      --branch "$(git branch --show-current)"
```
[the remark this answers is not in the document]
Comment (donovan-yohan) [reply skills/harness-brainstorm/SKILL.md:76]: Fixed. Added `.md` extension to `--design-doc` path. See 5f042f0.

File: plugins/harness/scripts/harness-read-state.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (donovan-yohan) [reply plugins/harness/scripts/harness-read-state.sh]: Fixed. Updated header comment to: "Returns {\"phase\": \"none\"} if file doesn't exist." See 5f042f0.

File: plugins/harness/commands/brainstorm.md [hunk 18]
```diff
@@ -52,15 +52,26 @@ Design through collaborative dialogue, saved as a versioned design document in t
...
    - Add to the "Current State" bullets if a new pattern was established
    - Add to the "Key Decisions" table if a non-trivial decision was made
 
+5.5. **Update run-state** (if `.harness/` runtime exists):
+    ```bash
+    HARNESS_DIR=$(bash ${CLAUDE_PLUGIN_ROOT}/scripts/harness-resolve-dir.sh --repo-root .)
+    [ -n "$HARNESS_DIR" ] && bash ${CLAUDE_PLUGIN_ROOT}/scripts/harness-update-state.sh \
+      --harness-dir "$HARNESS_DIR" \
+      --phase "brainstorm" \
+      --design-doc "docs/design-docs/{filename}" \
```
[the remark this answers is not in the document]
Comment (donovan-yohan) [reply plugins/harness/commands/brainstorm.md]: Fixed. Added `.md` extension to `--design-doc` path. See 5f042f0.

File: plugins/harness/.claude-plugin/plugin.json [hunk 19]
```diff
@@ -1,14 +1,15 @@
 {
   "name": "harness",
-  "version": "4.0.0",
-  "description": "3-tier documentation system with learning persistence, doc lifecycle management, and health scoring. Brainstorm, plan, orchestrate, review, reflect, complete with pr-review-toolkit agents and superpowers SDD delegation.",
+  "version": "5.0.1",
+  "description": "Self-improving documentation system with agent evolution, metrics-driven self-modification, and stateless command protocol. Brainstorm, plan, orchestrate, evolve, review, reflect, complete.",
```
[the remark this answers is not in the document]
Comment (donovan-yohan) [reply plugins/harness/.claude-plugin/plugin.json]: Fixed. Reordered to: "...review, reflect, evolve, complete." See 5f042f0.

File: plugins/harness/commands/evolve.md [hunk 20]
```diff
@@ -0,0 +1,237 @@
...
+
+16. For non-eligible proposals: leave status as `pending`. These will be surfaced in the PR description by `/harness:complete`.
+
+### Phase 5: Auto-Rollback Check
+
+17. Check if `$HARNESS_DIR/metrics/pre-change-snapshot.json` exists (written in Phase 4 step 14). If it doesn't exist, skip this phase — no changes were applied to roll back.
+
+18. Read `$HARNESS_DIR/memory/IMPROVEMENTS.md`. For each auto-applied change from previous sessions (not the current session):
+    - Compare current `$HARNESS_DIR/metrics/review-effectiveness.json` against `pre-change-snapshot.json`:
+      - False positive rate increased by more than 20%
```
[the remark this answers is not in the document]
Comment (donovan-yohan) [reply plugins/harness/commands/evolve.md:190]: Fixed. Scoped rollback evaluation to current-session proposals only, with a note explaining why (snapshot is overwritten each run). Applied to both `evolve.md` and `harness-evolve/SKILL.md`. See 5f042f0.

File: skills/harness-init/SKILL.md [hunk 21]
```diff
@@ -318,6 +318,55 @@ harness-init              # Full guided extraction
...
+    Determine the repo name for manifest:
+    ```bash
+    git remote get-url origin 2>/dev/null | sed 's|.*github.com[:/]||;s|\.git$||' || echo "local/$(basename $(pwd))"
+    ```
+
+    Run the scaffold script:
+    ```bash
+    bash ${CLAUDE_PLUGIN_ROOT}/scripts/harness-init-runtime.sh \
+      --harness-dir "$HARNESS_DIR" \
+      --repo-name "$REPO_NAME"
```
[the remark this answers is not in the document]
Comment (donovan-yohan) [reply skills/harness-init/SKILL.md:353]: Fixed. Changed to `REPO_NAME=$(git remote get-url ...)` to capture the output. Applied to both `init.md` and `harness-init/SKILL.md`. See 5f042f0.

File: plugins/harness/commands/init.md [hunk 22]
```diff
@@ -314,6 +314,55 @@ Transform a monolithic CLAUDE.md into a 3-tier progressive disclosure documentat
...
+    Determine the repo name for manifest:
+    ```bash
+    git remote get-url origin 2>/dev/null | sed 's|.*github.com[:/]||;s|\.git$||' || echo "local/$(basename $(pwd))"
+    ```
+
+    Run the scaffold script:
+    ```bash
+    bash ${CLAUDE_PLUGIN_ROOT}/scripts/harness-init-runtime.sh \
+      --harness-dir "$HARNESS_DIR" \
+      --repo-name "$REPO_NAME"
```
[the remark this answers is not in the document]
Comment (donovan-yohan) [reply plugins/harness/commands/init.md:349]: Fixed. Same as above — assigned command output to `REPO_NAME`. See 5f042f0.

File: skills/harness-evolve/SKILL.md [hunk 23]
```diff
@@ -0,0 +1,241 @@
...
+
+16. For non-eligible proposals: leave status as `pending`. These will be surfaced in the PR description by `harness-complete`.
+
+### Phase 5: Auto-Rollback Check
+
+17. Check if `$HARNESS_DIR/metrics/pre-change-snapshot.json` exists (written in Phase 4 step 14). If it doesn't exist, skip this phase — no changes were applied to roll back.
+
+18. Read `$HARNESS_DIR/memory/IMPROVEMENTS.md`. For each auto-applied change from previous sessions (not the current session):
+    - Compare current `$HARNESS_DIR/metrics/review-effectiveness.json` against `pre-change-snapshot.json`:
+      - False positive rate increased by more than 20%
```
[the remark this answers is not in the document]
Comment (donovan-yohan) [reply skills/harness-evolve/SKILL.md:194]: Fixed. Scoped rollback to current-session proposals only (matching `evolve.md`). See 5f042f0.

File: plugins/harness/scripts/harness-read-state.sh (hunk 1 above)
Comment (donovan-yohan) [reply plugins/harness/scripts/harness-read-state.sh]: Fixed. Updated header to match behavior: "Returns {\"phase\": \"none\"} if file doesn't exist." See 5f042f0.

File: plugins/harness/scripts/harness-read-state.sh (hunk 2 above)
Comment (donovan-yohan) [reply plugins/harness/scripts/harness-read-state.sh:18]: Fixed. Added JSON validation via `python3 -c "import json,sys; json.load(sys.stdin)"` before outputting. Returns error on corrupt JSON. See 5f042f0.

File: plugins/harness/scripts/harness-update-state.sh (hunk 3 above)
Comment (donovan-yohan) [reply plugins/harness/scripts/harness-update-state.sh]: Fixed. Changed comment to "Atomic write: write to temp file then rename..." in both `harness-update-state.sh` and `harness-write-metrics.sh`. See 5f042f0.

File: plugins/harness/scripts/harness-write-run.sh (hunk 4 above)
Comment (donovan-yohan) [reply plugins/harness/scripts/harness-write-run.sh]: Fixed. Added temp-file-then-rename pattern matching sibling scripts, with OSError handling and cleanup. See 5f042f0.

File: plugins/harness/scripts/harness-write-run.sh (hunk 5 above)
Comment (donovan-yohan) [reply plugins/harness/scripts/harness-write-run.sh]: Fixed. Replaced sed-based reformatting with two direct `date` calls from a shared epoch value (with macOS/Linux fallback). See 5f042f0.

File: plugins/harness/scripts/harness-write-metrics.sh (hunk 6 above)
Comment (donovan-yohan) [reply plugins/harness/scripts/harness-write-metrics.sh]: Fixed. Removed `disabled: False, disable_reason: None` from the `setdefault` dict. Updated runtime-spec.md NOTE to clarify both need to be added together when the feature is built. See 5f042f0.

File: plugins/harness/scripts/harness-init-runtime.sh (hunk 7 above)
Comment (donovan-yohan) [reply plugins/harness/scripts/harness-init-runtime.sh]: Fixed. Added `|| true` to the config.yaml backup line to prevent `set -e` abort when config.yaml doesn't exist. See 5f042f0.

File: plugins/harness/scripts/harness-init-runtime.sh (hunk 8 above)
Comment (donovan-yohan) [reply plugins/harness/scripts/harness-init-runtime.sh:121]: Fixed. Added `metrics/pre-change-snapshot.json` to the `.gitignore` template. See 5f042f0.

File: plugins/harness/scripts/harness-resolve-dir.sh (hunk 9 above)
Comment (donovan-yohan) [reply plugins/harness/scripts/harness-resolve-dir.sh:20]: Fixed. Added directory existence check with descriptive error before proceeding. See 5f042f0.

File: plugins/harness/scripts/harness-write-proposal.sh (hunk 10 above)
Comment (donovan-yohan) [reply plugins/harness/scripts/harness-write-proposal.sh:54]: Fixed. Added stderr warnings when explicitly-provided `--current-file`, `--proposed-file`, or `--reasoning-file` paths don't exist. See 5f042f0.

File: plugins/harness/commands/evolve.md (hunk 11 above)
Comment (donovan-yohan) [reply plugins/harness/commands/evolve.md]: Fixed. Changed to canonical `--learning-id`, `--recurrence`, `--prevented` parameters in both `evolve.md` and `harness-evolve/SKILL.md`. See 5f042f0.

File: plugins/harness/commands/review.md (hunk 12 above)
Comment (donovan-yohan) [reply plugins/harness/commands/review.md]: Fixed. Renumbered Report step from 23 to 28 in both `review.md` and `harness-review/SKILL.md`. See 5f042f0.
